### ISSUE TITLE
Dyno: track original names of extern types in addition to their linkage names.

### ISSUE THREAD

**Issue by DanilaFe** [chapel-lang/chapel#22711]
This is spinning off [a comment](https://github.com/chapel-lang/chapel/pull/22707#discussion_r1259008626) on #22707. 

Currently, extern types only store their linkage name. This is important because two types with the same linkage name are equivalent, even if they were defined in different places (this should match the intuition of declaring an external proc multiple times, though equality checking on procedures isn't a thing). However, we don't want to show linkage names to the user instead of the Chapel type name: if you write `extern "some_weird_t" type myType`, Chapel should report it as `myType`. Currently, extern types don't store the original name or the ID of where they're defined, and so we can't print the Chapel name.

The reason this is nontrivial that our type comparisons in canPass are based on pointer equality; however, if we were to store the ID or name of the variable that declared an opaque type in addition to the linkage name, two variables with distinct Chapel names but identical linkage names would not be the exact same type, and thus become incompatible. We don't want that. To address the issue, we probably want to do a deep comparison on types; however, this has implications for a fair chunk of our code (we normally use pointer equality; pointer equality is much faster than a deep comparison; do other types need deep comparisons too -- looking at you, CompositeType!). For the time being, the extern type PR left these questions unanswered, but they need to be answered eventually.

**Comment by mppf** [comment]
> our type comparisons in canPass are based on pointer equality ... do other types need deep comparisons too

I don't think we have to do an all-or-nothing thing here. We could introduce a method to help compare two types that only does more that pointer equality for extern types. Indeed, extern types are special in that we should be able to have several declarations of the same extern type and have them produce the same type. That wouldn't be true for a record, say.

> looking at you, CompositeType

Most CompositeTypes use a nominative typing strategy so it makes sense to do pointer comparison with pointers connected to the declaration. Tuples are structural but we should be handling that with queries that unique-ify so we get the same TupleType when we ask for a tuple with the same element types.

Of course, if we can have multiple `Type*` pointers for an extern type, that implies that our TupleType comparison will need to consider the elements, at least, if any of the tuple elements are extern types.

Another completely different approach is to represent extern types in a different way that makes these operations more natural. I'm not sure what that way would be, though.

Perhaps we could have the `Type*` for an extern type match the extern `cname` and be uniqueified based on that. I think that would work great for opaque extern types like `extern type memory_order`.

> However, we don't want to show linkage names to the user instead of the Chapel type name

Perhaps we could use a table to find, from a given scope, what the current user-facing name of an extern type that is in scope is? (Here I am expecting that extern types are only declared once in a given scope; we could warn if there are multiple type declarations using the same extern type name).

Also, this strategy doesn't seem smooth for things like `extern type wint_t = int(32)`, because in that kind of declaration, the Chapel type system should use `int(32)`, but somehow try to remember also the C type name. Historically, we have not remembered the C part by the time we are code generating, but that's something we'd like to improve (See issue #9324 and the issues linked there). I'm not sure it's possible to improve that without a lot of trouble, and it's less relevant than it used to be now that the LLVM backend is the default. In particular, while we could add some sort of `isTypeEquivalent` for dyno, we rely on the unique-ifying for lots of elements of function resolution & generics support, and it would be awkward to use `isTypeEquivalent` for those. From a language point of view, if I have `proc foo(arg) { }` and I call it with `var x: wint_t; foo(x);` and then with `var y: int(32); foo(y);`, should it use the same instantiation? I think it should, but if it does, we are necessarily forgetting information about the extern type involved.

Here is the direction I think I would head:
 * extern C types are created and uniquified based only on the C name
 * we use some queries to support computing the Chapel name from an extern C type, somehow. It could be scope dependent or maybe we can get away with just having a single Chapel name for a given C name & warn/error if multiple are defined.
